modules/dynamic/src/smithy4s/dynamic/DynamicModelCompiler.scala

```diff
@@ -0,0 +1,171 @@
+      val shapeId: ShapeId =
+        ShapeId.fromParts(ks.hintKey.namespace, ks.hintKey.name)
+      shapeId -> ks
+    }.toMap
+
+    def toHintAux[A](ks: KeyedSchema[A], node: Node): Option[Hint] = {
+      val documentRepr: Document = NodeToDoc(node)
+      val decoded: Option[A] =
+        Document.Decoder.fromSchema(ks.schema).decode(documentRepr).toOption
+      decoded.map { value =>
+        Hints.Binding(ks.hintKey, value)
+      }
+    }
+
+    def toHint(tr: traits.Trait): Option[Hint] = {
+      val id = tr.toShapeId()
+      val node = tr.toNode()
+      hintsMap.get(id).flatMap(toHintAux(_, node))
+    }
+
+    val visitor =
+      new CompileVisitor(model, schemaMap, endpointMap, serviceMap, toHint(_))
+    model.shapes().iterator().asScala.foreach(_.accept(visitor))
+    new DynamicModel(serviceMap.toMap, schemaMap.toMap)
+  }
+
+  class CompileVisitor(
+      model: Model,
+      schemaMap: MMap[ShapeId, Schema[DynData]],
+      endpointMap: MMap[ShapeId, DynamicEndpoint],
+      serviceMap: MMap[ShapeId, DynamicService],
+      toHint: traits.Trait => Option[Hint]
+  ) extends ShapeVisitor.Default[Unit] {
+
+    def allHints(shape: Shape): Seq[Hint] = {
+      shape
+        .getAllTraits()
+        .asScala
+        .values
+        .map(toHint)
+        .collect { case Some(h) =>
+          h
+        }
+        .toSeq
+    }
+
+    def allHints(shapeId: ShapeId): Seq[Hint] =
+      allHints(model.expectShape(shapeId))
+
+    def update[A](shape: Shape, schema: Schema[A]): Unit = {
+      schemaMap += (shape.getId -> schema
+        .withHints(allHints(shape): _*)
+        .withHints(Name(shape.getId.getName()))
+        .asInstanceOf[Schema[DynData]])
+    }
+
+    def getDefault(shape: Shape): Unit = ()
+
+    override def integerShape(shape: IntegerShape): Unit =
+      update(shape, int)
+
+    override def booleanShape(shape: BooleanShape): Unit =
+      update(shape, boolean)
+
+    override def stringShape(shape: StringShape): Unit =
+      update(shape, string)
+
+    override def listShape(shape: ListShape): Unit =
+      update(shape, list(suspend(schemaMap(shape.getMember.getTarget))))
+
+    override def setShape(shape: SetShape): Unit =
+      update(shape, set(suspend(schemaMap(shape.getMember.getTarget))))
+
+    override def operationShape(shape: OperationShape): Unit = {
+      def maybeSchema(maybeShapeId: Optional[ShapeId]): Schema[DynData] =
+        suspend(
+          maybeShapeId
+            .map[Schema[DynData]](id => schemaMap(id))
+            .orElseGet(() => unit.asInstanceOf[Schema[DynData]])
+        )
+
+      val ep = DynamicEndpoint(
+        shape.getId.getNamespace(),
+        shape.getId.getName(),
+        maybeSchema(shape.getInput()),
+        maybeSchema(shape.getOutput()),
+        Hints(allHints(shape): _*)
+      )
+      endpointMap += shape.getId() -> ep
+    }
+
+    override def serviceShape(shape: ServiceShape): Unit = {
+      val getEndpoints = () =>
```

Comment:
Soooo why is this lazy?

Reply:
For simplicity of implementation : we traverse the shapes only once. When we encounter a service, it's endpoints might not have been created yet.

That being said I could force evaluation at the end of the traversal, but in practice it shouldn't make a difference since the interpreters often precompile stuff on their end

Reply:
fair nuff